### PR TITLE
Use "Pump" to transfer http bodies when forwarding http-requests (and…

### PR DESCRIPTION
**Pull request by oli-h**
… their responses) to backends (or to some other Gateleen verticles).

This solves OutOfMemoryErrors with huge http-bodies in case of fast incoming- and a slow draining-stream

### CODE REVIEW

File: gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java [hunk 1]
```diff
@@ -200,11 +205,14 @@ private void handleRequest(final HttpServerRequest req, final Buffer bodyData, f
...
         if (bodyData == null) {
-            req.handler(data -> {
-                cReq.write(data);
-                loggingHandler.appendRequestPayload(data);
-            });
+            final LoggingWriteStream loggingWriteStream = new LoggingWriteStream(cReq, loggingHandler, true);
+            final Pump pump = Pump.pump(req, loggingWriteStream);
             req.endHandler(v -> cReq.end());
+            req.exceptionHandler(t -> {
+                LOG.error("Exception during forwarding - closing (forwarding) client connection", t);
```
[the remark this answers is not in the document]
Comment (oli-h) [reply gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java]: Found org.swisspush.gateleen.core.logging.RequestLogger?
But there it's not clear to me how to use this for error (resp. warn) logging.
Therefore I additionally found org.swisspush.gateleen.core.http.RequestLoggerFactory - which seems better suited for your requirement.

Please clarify. Thx

**Comment by oli-h** [comment]
I agree with the requested changes. Done.

**Comment by lbovet** [comment]
Did you test this in both ways?
I can reproduce a OOM with a command like
```curl localhost:7012/starwars.avi | hexdump -C
```
Where the routing goes to a local ngninx serving files (a 1.4G movie, in this case)

**Comment by oli-h** [comment]
Damn. We have a different execution path when there is an ETag present in the response.
This path does not (yet) use the Pump. Can you turn off ETags in ngninx and give it a try?

**Comment by lbovet** [comment]
Confirmed, this is because of the Etag...

**Comment by oli-h** [comment]
Why do we need ETag handling in Gateleen's Forwarder?
It should be the Job of the upstream server to hanlde ETags (and if-modified-since...) and answer "304 Not modified"  - and an empty body. If upstream server sends - for whatever reason - "200 OK" together with a body, even when ETag matches: how wan we 'Gateleen' decide to discard the body?

**Comment by lbovet** [comment]
I agree with your statement. This is weird to have this code snippet in the
first place.
I will dig in the history to see how and why it came there....

**Pull request merged**